search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceCaller.kt

```diff
@@ -11,28 +11,85 @@ import com.egm.stellio.search.csr.model.NGSILDWarning
 import com.egm.stellio.search.csr.model.RevalidationFailedWarning
 import com.egm.stellio.shared.model.CompactedEntity
 import com.egm.stellio.shared.queryparameter.QueryParameter
+import com.egm.stellio.shared.util.JsonUtils.deserializeAsList
+import com.egm.stellio.shared.util.JsonUtils.deserializeAsMap
+import com.egm.stellio.shared.util.RESULTS_COUNT_HEADER
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
-import org.springframework.core.codec.DecodingException
 import org.springframework.http.HttpHeaders
 import org.springframework.http.HttpMethod
 import org.springframework.http.HttpStatus
 import org.springframework.util.CollectionUtils
 import org.springframework.util.MultiValueMap
+import org.springframework.web.reactive.function.client.ClientResponse
 import org.springframework.web.reactive.function.client.WebClient
 import org.springframework.web.reactive.function.client.awaitBodyOrNull
 import org.springframework.web.reactive.function.client.awaitExchange
 import java.net.URI
 
+typealias QueryEntityResponse = Pair<List<CompactedEntity>, Int?>
+
 object ContextSourceCaller {
     val logger: Logger = LoggerFactory.getLogger(javaClass)
 
+    suspend fun retrieveContextSourceEntity(
+        httpHeaders: HttpHeaders,
+        csr: ContextSourceRegistration,
+        id: URI,
+        params: MultiValueMap<String, String>
+    ): Either<NGSILDWarning, CompactedEntity?> = either {
+        val path = "/ngsi-ld/v1/entities/$id"
+
+        return kotlin.runCatching {
+            getDistributedInformation(httpHeaders, csr, path, params).bind().first?.deserializeAsMap().right()
+        }.fold(
+            onSuccess = { it },
+            onFailure = { e ->
+                logger.warn("Error contacting CSR ${csr.id} at $path: ${e.message}")
+                logger.warn(e.stackTraceToString())
+                RevalidationFailedWarning(
+                    "${csr.id} at $path returned badly formed data message: \"${e.cause}:${e.message}\"",
+                    csr
+                ).left()
+            }
+        )
+    }
+
+    suspend fun queryContextSourceEntities(
+        httpHeaders: HttpHeaders,
+        csr: ContextSourceRegistration,
+        count: Boolean,
+        params: MultiValueMap<String, String>
+    ): Either<NGSILDWarning, QueryEntityResponse> = either {
+        val path = "/ngsi-ld/v1/entities"
+
+        return kotlin.runCatching {
```

Comment:
the `runCatching` is managed in `getDistributedInformation`, not anymore necessary here

Reply:
The deserialize as list can throw InvalidRequestException

Reply:
Maybe it is not necessary in getDistributedInformation instead.